airbyte-integrations/connectors/source-dv-360/source_dv_360/source.py

```diff
@@ -0,0 +1,153 @@
+        """
+        cred_json= config.get('credentials')
+        return Credentials(
+            token= cred_json.get('access_token'),
+            refresh_token= cred_json.get('refresh_token'),
+            token_uri= cred_json.get('token_uri'),
+            client_id= cred_json.get('client_id'),
+            client_secret= cred_json.get('client_secret')
+        )
+
+
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, any]:
+        """
+        Tests if the input configuration can be used to successfully connect to the integration
+            e.g: if a provided Stripe API token can be used to connect to the Stripe API.
+
+        :param logger: Logging object to display debug/info/error to the logs
+            (logs will not be accessible via airbyte UI if they are not passed to this logger)
+        :param config: Json object containing the configuration of this source, content of this json is as specified in
+        the properties of the spec.json file
+
+        :return: AirbyteConnectionStatus indicating a Success or Failure
+        """
+        try:
+            dbm_service= build('doubleclickbidmanager','v1.1', credentials=self.get_credentials(config))
+            request= dbm_service.queries().listqueries().execute()
+            return True, None
+        except Exception as err:
+            return False, f"Unable to connect to Google Ads API with the provided credentials - {repr(err)}"
+
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        """
+        :param config: The user-provided configuration as specified by the source's spec.
+        Any stream construction related operation should happen here.
+        :return: A list of the streams in this source connector.
+        """
+        args= dict(
+            credentials=self.get_credentials(config),
+            partner_id= config.get('partner_id'), 
+            start_date= config.get('start_date'), 
+            end_date= config.get('end_date'),
+            filters= config.get('filters')
+        ) 
+
+        streams= [
+            Reach(**args),
+            Standard(**args), 
+            AudienceComposition(**args),
+            Floodlight(**args),
+            UniqueReachAudience(**args),
+        ]
+        return streams
+        
+ 
+    def read(self, logger: AirbyteLogger, config: json, catalog: ConfiguredAirbyteCatalog, state: MutableMapping[str, Any]) -> Generator[AirbyteMessage, None, None]:
+        """
+        Returns a generator of the AirbyteMessages generated by reading the source with the given configuration,
+        catalog, and state.
+
+        :param logger: Logging object to display debug/info/error to the logs
+            (logs will not be accessible via airbyte UI if they are not passed to this logger)
+        :param config: Json object containing the configuration of this source, content of this json is as specified in
```

Comment:
do you really need to overwrite the read function?

Reply:
yes, because this API is special, not all fields from the same report are compatible with each other, therefore we should only use the fields specified in the configured catalog to read the data, instead of using the name of the stream and fetch all available fields from the schema.